sdk/core/Azure.Core/src/Shared/StatusCodeClassifier.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+#nullable enable
+
+namespace Azure.Core
+{
+    internal class StatusCodeClassifier : ResponseClassifier
+    {
+        private ulong[] _nonErrors = new ulong[10];
+
+        internal HttpMessageClassifier[]? MessageClassifiers { get; set; }
+
+        public StatusCodeClassifier(int[] nonErrors)
+        {
+            for (int i = 0; i < nonErrors.Length; i++)
+            {
+                AddClassifier(nonErrors[i], true);
+            }
+        }
+
+        private StatusCodeClassifier(ulong[] nonErrors)
+        {
+            Debug.Assert(nonErrors?.Length == 10);
+            Array.Copy(nonErrors, _nonErrors, _nonErrors.Length);
+        }
+
+        public override bool IsErrorResponse(HttpMessage message)
+        {
+            if (TryClassify(message, out var isError))
+            {
+                return isError;
+            }
+
+            if (MessageClassifiers != null)
+            {
+                for (int i = MessageClassifiers.Length; i >= 0; i--)
+                {
+                    if (MessageClassifiers[i].TryClassify(message, out isError))
+                    {
+                        return isError;
+                    }
+                }
+            }
+
+            return !IsNonError(message.Response.Status);
+        }
+
+        public virtual StatusCodeClassifier Clone()
+        {
+            return new StatusCodeClassifier(_nonErrors);
```

Comment:
I would do array copy here as opposed in the ctor. IMO it's easier to see what's going on.